### PR TITLE
Additional Task-related extension methods:

### PR DESCRIPTION
**Pull request by jskeet**
- WaitWithUnwrappedExceptions overloads to match Wait
- TaskCompletionSource.WithCancellationToken

### CODE REVIEW

File: Google.Api.Gax/TaskCompletionSourceExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+            GaxPreconditions.CheckNotNull(source, nameof(source));
+            if (!cancellationToken.CanBeCanceled)
+            {
+                return source.Task;
+            }
+            return Wait();
+
+            async Task<TResult> Wait()
+            {
+                using (var registration = cancellationToken.Register(() => source.TrySetCanceled()))
```
Comment (chrisdunelm) [inline Google.Api.Gax/TaskCompletionSourceExtensions.cs]: nit: `registration` local not required.

File: Google.Api.Gax/TaskExtensions.cs [hunk 2]
```diff
@@ -49,7 +50,98 @@ public static void WaitWithUnwrappedExceptions(this Task task)
...
+        }
 
+        /// <summary>
+        /// Synchronously waits for the given task to complete.
+        /// Any <see cref="AggregateException"/> thrown is unwrapped to the first inner exception.
+        /// </summary>
+        /// <param name="task">The task to wait for.</param>
+        /// <param name="millisecondsTimeout">The number of milliseconds to wait, or
+        /// -1 to wait indefinitely.</param>
+        public static void WaitWithUnwrappedExceptions(this Task task, int millisecondsTimeout)
```
Comment (chrisdunelm) [inline Google.Api.Gax/TaskExtensions.cs]: Do we use these overloads with `int` timeout parameters? If not, I'd rather we didn't include them here, as I don't really like `int`s being used as durations.

File: Google.Api.Gax.Tests/TaskCompletionSourceExtensionsTests.cs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+            var tcs = new TaskCompletionSource<int>();
+            var source = new CancellationTokenSource();
+            var task = tcs.WithCancellationToken(source.Token);
+
+            tcs.SetResult(5);
+            Assert.Equal(5, await task);
+        }
+
+        [Fact]
+        public async Task WithCancellationToken_Cancelled()
```
Comment (chrisdunelm) [inline Google.Api.Gax.Tests/TaskCompletionSourceExtensionsTests.cs]: Consider adding a test with an already-cancelled cancellation-token?

File: Google.Api.Gax/TaskCompletionSourceExtensions.cs (hunk 1 above)
Comment (jskeet) [reply Google.Api.Gax/TaskCompletionSourceExtensions.cs]: Good point. Will remove.

File: Google.Api.Gax.Tests/TaskCompletionSourceExtensionsTests.cs (hunk 3 above)
Comment (jskeet) [reply Google.Api.Gax.Tests/TaskCompletionSourceExtensionsTests.cs]: Sure.

File: Google.Api.Gax/TaskExtensions.cs (hunk 2 above)
Comment (jskeet) [reply Google.Api.Gax/TaskExtensions.cs]: Hmm. The point would be to make it trivial to go from a `Wait(anything)` to `WaitWithUnwrappedExtensions(anything)`, but I do see your point. Let's see what Amanda thinks :)

Comment (jskeet) [reply Google.Api.Gax/TaskExtensions.cs]: Note that if we do remove the milliseconds-based ones, we'd want to add a "TimeSpan and CancellationToken" one - which doesn't exist on Wait itself.

**Comment by chrisdunelm** [comment]
Agreed about adding "TimeSpan and CancellationToken" method.

File: Google.Api.Gax/TaskExtensions.cs [hunk 4]
```diff
@@ -49,7 +50,98 @@ public static void WaitWithUnwrappedExceptions(this Task task)
...
+        public static void WaitWithUnwrappedExceptions(this Task task, TimeSpan timeout)
+        {
+            try
+            {
+                task.Wait(timeout);
+            }
+            catch (AggregateException e)
+            {
+                // Unwrap the first exception, a bit like await would.
+                // It's very unlikely that we'd ever see an AggregateException without an inner exceptions,
```
Comment (amanda-tarafa) [inline Google.Api.Gax/TaskExtensions.cs]: super nit: without an inner exception (singular)

File: Google.Api.Gax/TaskExtensions.cs [hunk 5]
```diff
@@ -49,7 +50,98 @@ public static void WaitWithUnwrappedExceptions(this Task task)
...
+        public static void WaitWithUnwrappedExceptions(this Task task, int millisecondsTimeout)
+        {
+            try
+            {
+                task.Wait(millisecondsTimeout);
+            }
+            catch (AggregateException e)
+            {
+                // Unwrap the first exception, a bit like await would.
+                // It's very unlikely that we'd ever see an AggregateException without an inner exceptions,
```
Comment (amanda-tarafa) [inline Google.Api.Gax/TaskExtensions.cs]: nit: same here, inner exception (singular) and everywhere else.

File: Google.Api.Gax/TaskExtensions.cs (hunk 2 above)
Comment (amanda-tarafa) [reply Google.Api.Gax/TaskExtensions.cs]: Agree with `int` as duration being ugly, but it's already in `Wait`, and we are just wrapping `Wait` so, might make sense here. These are just helper methods after all. I'm more incline to leave it as is.

File: Google.Api.Gax/TaskExtensions.cs [hunk 6]
```diff
@@ -49,7 +50,98 @@ public static void WaitWithUnwrappedExceptions(this Task task)
...
+        }
+
+        /// <summary>
+        /// Synchronously waits for the given task to complete.
+        /// Any <see cref="AggregateException"/> thrown is unwrapped to the first inner exception.
+        /// </summary>
+        /// <param name="task">The task to wait for.</param>
+        /// <param name="timeout">A TimeSpan that represents the number of milliseconds to wait, or
+        /// -1 milliseconds to wait indefinitely.</param>
+        public static void WaitWithUnwrappedExceptions(this Task task, TimeSpan timeout)
```
Comment (amanda-tarafa) [inline Google.Api.Gax/TaskExtensions.cs]: name wise, I would prefer `WaitAndUnwrapExceptions`, but not set on it.

File: Google.Api.Gax/TaskExtensions.cs (hunk 4 above)
Comment (jskeet) [reply Google.Api.Gax/TaskExtensions.cs]: Will fix this everywhere.

File: Google.Api.Gax/TaskExtensions.cs (hunk 6 above)
Comment (jskeet) [reply Google.Api.Gax/TaskExtensions.cs]: We've already got the original WaitWithUnwrappedExceptions method, so we can't really change that now.

**Comment by jskeet** [comment]
Will chat with Chris about the millisecond part...

**Comment by jskeet** [comment]
Discussed with Chris - repushed with the comment fixes; will merge on green.

**Pull request merged**